src/Locks.ts

```diff
@@ -0,0 +1,237 @@
+    super();
+  }
+
+  get(id: string): Lock | undefined {
+    for (const member of this.space.getMembers()) {
+      if (!member.locks) {
+        continue;
+      }
+      const request = member.locks.get(id);
+      if (request && request.status === LockStatus.LOCKED) {
+        return { request, member };
+      }
+    }
+  }
+
+  async acquire(id: string, opts?: LockOptions): Promise<LockRequest> {
+    const self = this.space.getSelf();
+    if (!self) {
+      throw new Error('Must enter a space before acquiring a lock');
+    }
+
+    // check there isn't an existing PENDING or LOCKED request for the current
+    // member, since we do not support nested locks
+    let req = self.locks && self.locks.get(id);
+    if (req && req.status !== LockStatus.UNLOCKED) {
+      throw ERR_LOCK_REQUEST_EXISTS;
+    }
+
+    // initialise a new PENDING request
+    req = {
+      id,
+      status: LockStatus.PENDING,
+      timestamp: Date.now(),
+    };
+    if (opts) {
+      req.attributes = opts.attributes;
+    }
+    if (!self.locks) {
+      self.locks = new Map();
+    }
+    self.locks.set(id, req);
+
+    // reflect the change in the member's presence data
+    await this.space.updateSelf(self);
+
+    return req;
+  }
+
+  async release(id: string): Promise<void> {
+    const self = this.space.getSelf();
+    if (!self) {
+      throw new Error('Must enter a space before acquiring a lock');
+    }
+
+    if (!self.locks) {
+      return;
+    }
+    self.locks.delete(id);
+
+    await this.space.updateSelf(self);
+  }
+
+  subscribe<K extends EventKey<LockEventMap>>(
+    listenerOrEvents?: K | K[] | EventListener<LockEventMap[K]>,
+    listener?: EventListener<LockEventMap[K]>,
+  ) {
+    try {
+      super.on(listenerOrEvents, listener);
+    } catch (e: unknown) {
+      if (e instanceof InvalidArgumentError) {
+        throw new InvalidArgumentError(
+          'Locks.subscribe(): Invalid arguments: ' + inspect([listenerOrEvents, listener]),
+        );
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  unsubscribe<K extends EventKey<LockEventMap>>(
+    listenerOrEvents?: K | K[] | EventListener<LockEventMap[K]>,
+    listener?: EventListener<LockEventMap[K]>,
+  ) {
+    try {
+      super.off(listenerOrEvents, listener);
+    } catch (e: unknown) {
+      if (e instanceof InvalidArgumentError) {
+        throw new InvalidArgumentError(
+          'Locks.unsubscribe(): Invalid arguments: ' + inspect([listenerOrEvents, listener]),
+        );
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  processPresenceMessage(message: Types.PresenceMessage) {
+    const member = this.space.getMemberFromConnection(message.connectionId);
+    if (!member) {
+      return;
+    }
+
+    if (!message.extras || !message.extras.locks || !Array.isArray(message.extras.locks)) {
+      // there are no locks in presence, so release any existing locks for the
+      // member
+      if (member.locks && member.locks.size > 0) {
```

Comment:
`member.locks instanceOf Map` might be more explicit

Reply:
Now that `member.locks` is always defined, I've removed this check.